### PR TITLE
frontend: refactor account into api/account

### PR DESCRIPTION
**Pull request by thisconnect**
Some components contain api interfaces, that got wildly imported

all over and grew organically. This lead to incomplete and duplicate
interfaces and is hard to maintain.

Centralizing all api interfaces and promises into web/src/api helps
reducing duplicates and separates api from ui components.
Additionally promise return values from api all come typed.

This is the first step and collects all account interfaces and api
calls into api/account.ts. Next step is to refactor all components
to import the interfaces and api calls from api/account.

### CODE REVIEW

File: frontends/web/src/api/account.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+export interface INoteTx {
+    internalTxID: string;
+    note: string;
+}
+
+export const postNotesTx = (code: string, {
+    internalTxID,
+    note,
+}: INoteTx): Promise<any> => {
```
Comment (thisconnect) [inline frontends/web/src/api/account.ts:126]: TODO fix <any>

File: frontends/web/src/api/account.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
+/**
+ * Copyright 2020 Shift Crypto AG
```
Comment (benma) [inline frontends/web/src/api/account.ts:2]: 2021 :P

**Review by benma — APPROVED** [review]
concept ACK, but I didn't look at the diff in detail. Lmk if you want me to.

in the future, it might nice to put all the api calls into a "class":

let accountApi = new AccountApi(accountCode)
accountApi.transactionList()
accountApi.export()
accountApi.canVerifyXPub()

etc

**Comment by thisconnect** [comment]
> in the future, it might nice to put all the api calls into a "class":

why?

I'd prefer to just keep it simple,  in the future with the new react hooks api, we shouldn't need to use class ever again.

**Comment by benma** [comment]
It was also about moving the account code in the constructor and not repeating that in every call. Not so important though, was just a thought.

I do like the `import accountApi from './api/account';` style!

File: frontends/web/src/api/account.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+export const sendTx = (code: string): Promise<ISendTx> => {
+    return apiPost(`account/${code}/sendtx`);
+};
+
+export type FeeTargetCode = 'custom' | 'low' | 'economy' | 'normal' | 'high';
+
+export interface IProposeTxData {
+    address?: string;
+    amount?: number;
+    // data?: string;
```
Comment (x1ddos) [inline frontends/web/src/api/account.ts:172]: data? what?

File: frontends/web/src/routes/account/account.tsx [hunk 4]
```diff
@@ -16,23 +16,32 @@
...
 import { Header } from '../../components/layout';
 import { Spinner } from '../../components/spinner/Spinner';
 import Status from '../../components/status/status';
-import { TransactionInterface } from '../../components/transactions/transaction';
 import { Transactions } from '../../components/transactions/transactions';
 import { load } from '../../decorators/load';
 import { subscribe } from '../../decorators/subscribe';
 import { translate, TranslateProps } from '../../decorators/translate';
-import { apiGet, apiPost } from '../../utils/request';
+import { apiGet } from '../../utils/request';
```
[the remark this answers is not in the document]
Comment (x1ddos) [reply frontends/web/src/routes/account/account.tsx:31]: if it is used... why not refactor the rest?

File: frontends/web/src/routes/account/account.tsx [hunk 5]
```diff
@@ -231,12 +225,12 @@ class Account extends Component<Props, State> {
         if (this.state.fatalError) {
             return;
         }
-        apiPost(`account/${this.props.code}/export`).then(exported => {
+        accountApi.exportAccount(this.props.code).then(exported => {
             this.setState({ exported });
-        });
+        }).catch(console.error);
```
Comment (x1ddos) [inline frontends/web/src/routes/account/account.tsx:230]: Would love to see these errors surfaced to the user, in the future.

**Review by x1ddos — APPROVED** [review]
LGTM

File: frontends/web/src/routes/account/account.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (thisconnect) [reply frontends/web/src/routes/account/account.tsx:31]: still used will be gone once we have `api/devices.ts`, do you want `api/devices` in this PR?

File: frontends/web/src/api/account.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+export const sendTx = (code: string): Promise<undefined | SendTxInterface> => {
+	return apiPost(`account/${code}/sendtx`);
+};
+
+export type FeeTargetCode = 'custom' | 'low' | 'economy' | 'normal' | 'high';
+
+export interface proposeTxData {
+	address?: string;
+	amount?: number;
+	// data?: string;
```
[the remark this answers is not in the document]
Comment (thisconnect) [reply frontends/web/src/api/account.ts]: That is something about ETH signing data, see https://github.com/digitalbitbox/bitbox-wallet-app/commit/983dbc8

File: frontends/web/src/api/account.ts (hunk 3 above)
Comment (thisconnect) [reply frontends/web/src/api/account.ts:172]: https://github.com/digitalbitbox/bitbox-wallet-app/commit/983dbc8

File: frontends/web/src/routes/account/account.tsx (hunk 5 above)
Comment (thisconnect) [reply frontends/web/src/routes/account/account.tsx:230]: yes 👍

File: frontends/web/src/api/account.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+export const canVerifyXPub = (code: string): Promise<number[]> => {
+	return apiGet(`account/${code}/can-verify-extended-public-key`);
+};
+
+export interface ReceiveAddress {
+    addressID: string;
+    address: string;
+}
+
+// export const verifyAddress = (code: string, addressID: string): Promise<> => {
```
[the remark this answers is not in the document]
Comment (thisconnect) [reply frontends/web/src/api/account.ts:147]: Will be needed, just didn't check what it returns, it is used in frontends/web/src/routes/account/info/signingconfiguration.tsx

I can convert signingconfiguration.tsx or everything if you want, but was afraid that the PR is too big. That is why at the moment only account.tsx and accountsummary.tsx are using `api/account` + every component that exported some itnerface used by account.tsx or accountsummary.tsx.

**Comment by x1ddos** [comment]
Unfortunately, that commit has no explanation.

 > Commented as we might bring it back soon; will remove completely if
not.

Seems to me the "if not" moment has come? unless "soon" meant several years.

**Comment by x1ddos** [comment]
No, it's fine. But I was wondering why not move apiGet and apiPost to 
the api/ namespace as well. To those two would've been the first candidates.

**Comment by thisconnect** [comment]
> No, it's fine. But I was wondering why not move apiGet and apiPost to
the api/ namespace as well.

Once everything is moved to api,, apiGet and apiPost will only be used inside api, so no need to expose it

**Comment by x1ddos** [comment]
Ok!

**Pull request merged**